File: tests/slicing/select_by_annot.c
/* run.config
   EXECNOW: make -s tests/slicing/select_by_annot.opt
   CMD: tests/slicing/select_by_annot.opt
   OPT: -deps -lib-entry -main main -journal-disable
   CMD: bin/toplevel.opt
   OPT: -check -deps -lib-entry -main main -slice-pragma main -journal-disable -then-on 'Slicing export' -print
   OPT: -check -deps -lib-entry -main main -slice-assert main -journal-disable -then-on 'Slicing export' -print
   OPT: -check -deps -lib-entry -main main -slice-pragma modifS -no-slice-callers -journal-disable -then-on 'Slicing export' -print
   OPT: -check -deps -lib-entry -main main -slice-pragma f1 -no-slice-callers -journal-disable -then-on 'Slicing export' -print
   OPT: -check -deps -lib-entry -main main -slice-pragma f2 -no-slice-callers -journal-disable -then-on 'Slicing export' -print
   OPT: -check -deps -lib-entry -main main -slice-pragma f3 -no-slice-callers -journal-disable -then-on 'Slicing export' -print
   OPT: -check -deps -lib-entry -main main -slice-pragma f4 -no-slice-callers -journal-disable -then-on 'Slicing export' -print
   OPT: -check -deps -lib-entry -main main -slice-pragma f5 -no-slice-callers -journal-disable -then-on 'Slicing export' -print
   OPT: -check -deps -lib-entry -main main -slice-pragma f6 -no-slice-callers -journal-disable -then-on 'Slicing export' -print
   OPT: -check -deps -lib-entry -main main -slice-pragma f7 -no-slice-callers -journal-disable -then-on 'Slicing export' -print
   OPT: -check -deps -lib-entry -main main -slice-loop-inv f8 -no-slice-callers -journal-disable -then-on 'Slicing export' -print
   OPT: -check -deps -lib-entry -main main -slice-pragma f8 -no-slice-callers -journal-disable -then-on 'Slicing export' -print
   OPT: -check -deps -lib-entry -main main -slice-assert f8 -no-slice-callers -journal-disable -then-on 'Slicing export' -print
   OPT: -check -deps -lib-entry -main main -slice-pragma f9 -no-slice-callers -journal-disable -then-on 'Slicing export' -print


*/
struct Tstr { int a; int b; } S;
int Sa ;

int f1(int cond) {
  int * p = &S.a ;
  if (cond) {
    //@ assert (cond != 0);
    Sa = *p ;
    }
  //@slice pragma expr *p;
  return Sa ;
}

int f2(int cond) {
  int * p = &S.a ;
  if (cond)
    //@ assert (cond != 0);
    Sa = *p ;
  //@slice pragma expr S.a;
  return Sa ;
}

int f3(int cond) {
  int * p = &S.a ;
  if (cond) {
    //@ slice pragma ctrl;
    Sa = *p ;
    }
  return Sa ;
}

int f4(int cond) {
  int * p = &S.a ;
  if (cond) {
    //@ slice pragma stmt;
    Sa = *p ;
    }
  return Sa ;
}

int f5(int cond) {
  int * p = &S.a ;
  if (cond) {
    //@ slice pragma expr 1;
    Sa = *p ;
    }
  return Sa ;
}

int f6(int cond) {
  int * p = &S.a ;
  //@ slice pragma stmt;
  if (cond) {
    Sa = *p ;
    Sa ++ ;
    }
  return Sa ;
}

int f7(int cond) {
  int * p = &S.a ;
  if (cond)
  //@ slice pragma stmt;
    {
      Sa = *p ;
      Sa ++ ;
    }
  return Sa ;
}


int f8(int cond) {
  int * p = &S.a ;
  //
  /*@ loop invariant cond >= 0 ;
    loop variant cond ; */
  while (cond)
    { //@ assert  cond <= \at(cond,Pre) ;
      //  assert S.a + cond == \at(S.a + cond,Pre) ;
      Sa = *p ;
      //@ slice pragma stmt;
      S.a ++ ;
      cond--;
    }
  return Sa ;
}

int X9, Y9, Z9 ;
void f9(int c1, int c2) {
  if (c1 > c2)
    goto L;
  c1 = c2 ;
  //@ slice pragma stmt;
  {L: X9 = c1 ;}
  Y9 = Z9 ;
  Z9 = c2 ;
}

void modifS (int a, int b) {
  S.a += a;
  S.b -= b;
  //@slice pragma expr S.a;
}
int new_int (void);
  int d;
int main (void) {
  int a = 0;
  int b = 0;
  int c = 0;
  if (d > 0) {
    //@ assert (b == 0);
    a = 1;
    }
  //@ slice pragma expr a+b;
  int x = a+b+c+d;
  modifS (a, b);
  // assert (d>0 => a == 1) && (!(d>0) => a==0);
  d = new_int ();
  f1(d) ;
  f2(d) ;
  f3(d) ;
  f4(d) ;
  f5(d) ;
  f6(d) ;
  f7(d) ;
  f8(d) ;
  f9(d,a) ;
  return x;
}

File: tests/slicing/select_by_annot.ml
(* ledit bin/toplevel.top -deps tests/slicing/select_by_annot.c \
   < tests/slicing/select_by_annot.ml
*)


open LibSelect;;

let main _ =
  Slicing.Api.Project.reset_slicing ();
  let pretty_pdg fmt kf = !Db.Pdg.pretty fmt (!Db.Pdg.get kf) in
  let add_annot kf =
    let mark = Slicing.Api.Mark.make ~data:true ~addr:false ~ctrl:false in
    let select = Slicing.Api.Select.empty_selects in
    let select = Slicing.Api.Select.select_func_annots select mark
      ~spare:true ~threat:false ~user_assert:false ~slicing_pragma:true
      ~loop_inv:true ~loop_var:true kf
    in
    Slicing.Api.Request.add_persistent_selection select
      (*Slicing.Api.Request.read_annotations project kf_main ;;*)
  in
  let kf_main = Globals.Functions.find_def_by_name "main" in
  add_annot kf_main;
  Format.printf "@[%a@]@\n" pretty_pdg kf_main;

  let kf_modifS = Globals.Functions.find_def_by_name "modifS" in
  (*Slicing.ApiRequest.read_annotations project kf_modifS ;;*)
  add_annot kf_modifS;
  Format.printf "@[%a@]@\n" pretty_pdg kf_modifS;

  Slicing.Api.Request.pretty Format.std_formatter;
  Slicing.Api.Request.apply_all_internal ();

  Slicing.Api.Project.pretty Format.std_formatter;
  extract_and_print ();

  (** create another slice for "main" to check if it also contains the previous
      * selection. *)

  let ff = Slicing.Api.Slice.create kf_main in
  let select = LibSelect.select_data "b" kf_main in
  Slicing.Api.Request.add_slice_selection_internal ff select;

  Slicing.Api.Request.apply_all_internal ();

  Slicing.Api.Project.pretty Format.std_formatter;
  extract_and_print ()

let () = Db.Main.extend main
